Refactor(*): JPA 관련 기능을 Domain 모듈로 이동 (SWM-63)

morenow98 opened the pull request:
## PR 변경된 내용
### 예외 처리 시 메시지도 넘기고, 해당 메시지로 예외를 처리하면 다음과 같은 문제가 발생해 구현 변경
    1. 에러 코드 클래스의 메시지는 쓰이지 않음
    2. 예외를 던질 때 넘기는 메시지를 관리하기 어려움
- JPA가 Domain 모듈에 이동되면서 Business 로직 예외와 영속성 관련 로직 예외를 한 패키지 안에서 관리해야 해서 두 예외를 분리해서 구현해 예외 발생 관리를 용이하게 함 -> 리뷰 결과 분리 구현하지 않고 합치는 것으로 결정
- ErrorCode 클래스에는 BusinessException, PersistenceException 관련 에러코드를 함께 관리

### JPA 의존성을 Infrastructure 모듈에서 Domain 모듈로 이동
- Infrastructure 모듈에 있던 매핑관련 클래스들 모두 삭제
- Infrastructure 모듈에는 일단 클래스 파일이 없고 yml 설정파일만 존재
- Domain 모듈의 Entity를 POJO에서 JPA가 관리하는 Entity로 변경

### JpaRepository interface 를 api 모듈에서 그대로 사용한다면 에러가 발생해 repository를 구현한 구현체를 둠으로써 에러를 해결하고 영속성 관련 예외처리도 구현
![image](https://github.com/Jabiseo/Jabiseo-Backend/assets/112200350/d762c936-5b31-49b1-96ca-9448037ce258)
![image](https://github.com/Jabiseo/Jabiseo-Backend/assets/112200350/d0fa46d5-0db1-4735-a6f8-fc4c8f456e25)
```
위 사진처럼 jpa를 implementation 말고 api를 통해 의존한다면 api 모듈에서 JpaRepository interface를 참조할 수 있다.
하지만 이렇게 되면 api 모듈이 JPA와 연관되어 멀티 모듈로서 의미가 떨어진다.
따라서 JpaRepository를 구현한 구현체를 둠으로써 implementation으로 의존 가능하도록 했다.
또한 예외처리도 구현했다.
```

## 추가 내용
- Infrastructure 모듈에서 Domain 모듈로 JPA가 이동함에 따라, Domain 모듈 안에서 패키지별로 DIP를 적용했다.
### yml 파일이 여러 모듈에 있어야 하는 상황의 해결
- Domain 모듈과 Infrastructure 모듈에 yml 파일이 중복해서 있어야 하는 상황.
- SpringBoot 가 돌아갈 때 특정한 순서로 yml 파일에 접근하게 되고, 먼저 접근한 yml 파일에 설정정보가 부족하다면 에러 발생.
- 따라서 SpringBoot 의존성이 있는 모듈에서 다음과 같은 코드 삽입
`System.setProperty("spring.config.name", "application, application-infrastructure");`
- 메인함수에 이 코드를 추가. 메인함수에 이 코드가 위치하는 게 올바르다고 생각하지는 않지만, 추후 리팩토링 예정.
### yml 파일간의 우선순위 문제 해결
- application-infrastructure.yml 파일이 먼저 적용되어 spring.datasource가 설정되어, 뒤에 오는 application-domain.yml 파일의 spring.jpa 설정값들이 무시되는 현상 발생
- domain 모듈의 설정파일은 application-domain.yml 이었지만, 우선순위를 높이기 위해 application.yml 로 이름 변경
- 참고: https://velog.io/@yeonjoo913/Spring-%EB%A9%80%ED%8B%B0-%EB%AA%A8%EB%93%88-%EC%9D%98%EC%A1%B4%EC%84%B1-%EA%B4%80%EB%A6%ACversionyml

## 참조
Jira Issue: SWM-63

morenow98 committed:
Refactor(*): 예외 처리 시 메시지는 넘기지 않고 에러 코드만 넘기는 것으로 변경

- 예외 처리 시 메시지도 넘기고, 해당 메시지로 예외를 처리하면 다음과 같은 문제가 발생해 구현 변경
1. 에러 코드 클래스의 메시지는 쓰이지 않음
2. 예외를 던질 때 넘기는 메시지를 관리하기 어려움

morenow98 committed:
Refactor(*): Domain 모듈에 BusinessException, PersistenceException 분리 구현

- JPA가 Domain 모듈에 이동되면서 Business 로직 예외와 영속성 관련 로직 예외를 한 패키지 안에서 관리해야 함.
- 두 예외를 분리해서 구현해서 예외 발생 관리를 용이하게 함
- ErrorCode 클래스에는 BusinessException, PersistenceException 관련 에러코드를 함께 관리

morenow98 committed:
Refactor(*): JPA를 Domain 모듈로 이동

- JPA 의존성을 Infrastructure 모듈에서 Domain 모듈로 이동
- Infrastructure 모듈에 있던 매핑관련 클래스들 모두 삭제
- Infrastructure 모듈에는 일단 클래스 파일이 없고 yml 설정파일만 존재
- Domain 모듈의 Entity를 POJO에서 JPA가 관리하는 Entity로 변경
- JpaRepository interface 를 api 모듈에서 그대로 사용한다면 에러가 발생해 repository를 구현한 구현체를 둠으로써 에러를 해결하고 영속성 관련 예외처리도 구현

morenow98 committed:
Chore(*): application.yml 파일을 gitignore 파일에서 제외하고 개인정보를 외부 주입으로 관리하는 것으로 변경

InHyeok-J commented on jabiseo-domain/src/main/java/com/jabiseo/certificate/exception/CertificatePersistenceException.java:
```diff
@@ -0,0 +1,12 @@
+import com.jabiseo.exception.BusinessException;
+import com.jabiseo.exception.ErrorCode;
+
+public class CertificatePersistenceException extends BusinessException {
+
+    public CertificatePersistenceException(ErrorCode errorCode) {
+        super(errorCode);
+    }
+
```
Domain안에 PersistenceException을 만든 이유가 있을까요 ? 
그리고 이건 BusineeeException을 상속하네요

JPA-> 결국 데이터베이스를 추상화시킨 기술이고 해당 로직에서 발생한 Exception은 BusineeException으로 합쳐도 된다고 생각합니다.
PersistenceException은 데이터 저장과 관련한 기술적인 Exception에 넣는건 어떨까요(Infra쪽)

morenow98 replied:
위 코드는 에러네요...

JPA 관련 에러와 비즈니스 로직 관련 에러를 분리해서 관리하고 싶어서 고민끝에 분리를 했습니다만,,
BusinessException 으로 합쳐도 좋을 것 같습니다.

BusinessException을 Domain 모듈에 유지시키고 PersistenceException은 Infrastructure 모듈로 이전할게요.

morenow98 replied:
그리고 위에서 말한 api 모듈이 infra 모듈을 의존하지 않는 것은,
PersistenceException이 infra 모듈에 위치하게 되면 RestControllerAdvice로 핸들링해야 하기 때문에 의존해야 해서 안될 것 같습니다!

InHyeok-J commented on jabiseo-domain/src/main/java/com/jabiseo/exception/BusinessException.java:
```diff
@@ -7,8 +7,7 @@ public class BusinessException extends RuntimeException {
     private final ErrorCode errorCode;
 
-    public BusinessException(String message, ErrorCode errorCode) {
-        super(message);
+    public BusinessException(ErrorCode errorCode) {
         this.errorCode = errorCode;
     }
```
에러 코드 관련해서 무슨 말인지는 알겠습니다.(PR내용)
근데 추 후 디버깅 용이나 RuntimeException을 핸들링을 해야 할 수도 있으니 super에 message를 보내주는건 어떨까요
super(errrorCode.getMessage()) 이렇게요

morenow98 replied:
아주 좋은 생각입니다 😁 바로 수정했습니다

InHyeok-J commented on jabiseo-infrastructure/src/main/resources/application.yml:
```diff
@@ -0,0 +1,15 @@
+    username: ${DB_USERNAME}
+    password: ${DB_PASSWORD}
+
+  jpa:
+    open-in-view: false
+    hibernate:
+      ddl-auto: ${DDL_AUTO}
+    properties:
+      dialect: org.hibernate.dialect.MySQL8InnoDBDialect
+    defer-datasource-initialization: true
+  sql.init.mode: always
```
build gradle은 이동이 됐는데 설정 파일에서 JPA설정은 왜 여기서 하나요??

morenow98 replied:
그러네요..! Database 설정이라고 착각했나봅니다... 수정했습니다

morenow98 replied:
여기서 문제가 발생해 본문에 추가했습니다. 여러 모듈에서 yml 파일을 사용할 때는 SpringBoot가 돌아가는 모듈에서 다음과 같은 설정을 해야 합니다.
`System.setProperty("spring.config.name", "application-infrastructure,application");`
메인함수에 이 코드를 추가했습니다. 메인함수에 이 코드가 위치하는 게 올바르다고 생각하지는 않지만, 추후 리팩토링 예정입니다.

InHyeok-J commented on jabiseo-infrastructure/src/main/resources/application.yml:
```diff
@@ -0,0 +1,15 @@
+spring:
+  datasource:
+    driver-class-name: com.mysql.cj.jdbc.Driver
+    url: ${DB_ENDPOINT}
+    username: ${DB_USERNAME}
```
DB_ENDPOINT -> DB_IP/DB_NAME/DP_PORT <- 이렇게 포함된 URL로 수정 가능할까요

morenow98 replied:
127.0.0.1 과 같이 IP형식이 아니고 다른 형식일 수도 있어서 DB_IP는 적절하지 않을 수도 있을 것 같다는 생각이 드네요.
그리고 만약 serverTimezone=Asia/Seoul 같은 설정을 붙여야 하는 경우가 있어도 문제일 것 같아요.

InHyeok-J replied:
정정할게요 DB_HOST 네요 그 외에 중복되지 않는건 변수처리하지 않고 상수로 박는게 깔끔해보여서 말씀드렸습니다

morenow98 replied:
아 좋습니다
`url: ${DB_HOST}/${DB_PORT}/${DB_NAME}/?serverTimezone=UTC&characterEncoding=UTF-8`
이렇게 변경할게요

morenow98 replied:
url: jdbc:mysql://${DB_HOST}:${DB_PORT}/${DB_NAME}?useSSL=false&useUnicode=true&serverTimezone=Asia/Seoul
와 같이 변경했습니다.

그리고 상의 결과 useSSL=false 설정의 경우 개발 환경에서 사용하고 배포 환경 등에서는 분리하겠습니다. 아래 URL과 같이 할 계획입니다.
https://github.com/slate-movte/slate-back/blob/develop/src/main/resources/application.yml

InHyeok-J reviewed (commented):
Domain의 규모가 Domain 객체와 데이터 접근 로직이 혼합된것 같습니다.
지금은 괜찮을것 같은데 나중에 데이터 접근 구현부와 도메인 로직 구현부의 혼동이 될 수도 있다는 생각이 들어요

아래 처럼 한번 패키기 구조를 분기 처리하고 현재 develop 기준의 inferface<-> impl을 유지하는건 어떤가요?
```
Domain-Module
/certificate
   /doamin
    /jpa
    /exception
```

그리고 여기 변경사항이 안나와있어서
build.gradle 기준 api에서 infra를 의존할 필요가 없어보입니다.( 직접 의존하는 클래스 X)
혹시 문제가 안된다면 삭제해도 될거같아요

morenow98 committed:
Refactor(*): Exception을 Business와 Persistence로 분리 구현하는 것에서 합치는 것으로 변경

morenow98 committed:
Refactor(*): Exception 생성자에 message도 포함시키도록 변경

- 추후 디버깅 등 RuntimeException을 핸들링하는 경우를 위해 변경

morenow98 committed:
Chore(*): JPA 관련 설정들을 infrastructure 모듈에서 domain 모듈로 변경

morenow98 committed:
Refactor(*): 설정 파일을 분리하기 위해 이름을 변경하고 설정하는 코드 추가

- 다른 모듈들에 yml 파일이 흩어져 있으면 오류 발생
- System.setProperty() 메소드를 통해 SpringBoot 의존성이 있는 모듈에서 다른 모듈의 yml 파일에 접근

morenow98 committed:
Refactor(*): DB_ENDPOINT 외부 설정값을 DB_HOST, DB_PORT, DB_NAME으로 분리

morenow98 committed:
Feat(*): Domain 모듈에서 패키지를 분리하고 패키지 간의 DIP 적용

- repository 패키지가 domain 패키지를 의존하도록 DIP 적용
- domain 패키지에 인터페이스를, repository 패키지에 구현체를 위치

morenow98 committed:
Fix(*): 다중 설정파일로 전환했지만, 실제 JPA 설정이 적용되지 않는 문제 해결

- infra(structure), domain 모듈의 설정파일이 모두 적용되는 것은 맞음
- 하지만 infra 모듈의 설정파일 먼저 적용되면서 그 뒤에 적용되는 domain 모듈의 설정파일이 무시되는 현상 발생
- domain 모듈의 설정 파일의 우선순위를 높이기 위해 application-domain.yml 에서 application.yml 파일로 파일 이름 변경
